### ISSUE TITLE
Since 1.97.0, `clippy::question_mark` fires on code matching on `&Option`

### ISSUE THREAD

**Issue by ChayimFriedman2** [rust-lang/rust-clippy#17386]
### Summary

Since 1.97.0 Clippy fires `clippy::question_mark` on code that handles `&Option`, while it shouldn't.

### Lint Name

question_mark

### Reproducer

I tried this code:

```rust
fn foo() -> &'static Option<i32> {
    &None
}

fn bar() -> Option<()> {
    match foo() {
        Some(it) => {
            dbg!(it);
        }
        None => return None,
    }
    Some(())
}
```

I saw this happen:

```none
warning: this `match` expression can be replaced with `?`
  --> src/lib.rs:6:5
   |
 6 | /     match foo() {
 7 | |         Some(it) => {
 8 | |             dbg!(it);
...  |
11 | |     }
   | |_____^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.97.0/index.html#question_mark
   = note: `#[warn(clippy::question_mark)]` on by default
help: try instead
   |
 6 ~     {
 7 +         let it = foo()?;
 8 +         dbg!(it);
 9 +     }
   |
```

I expected to see this happen:

The lint should not fire. The suggested fix does not compile, you cannot use `?` on `&Option`.

### Version

```text

```

**Comment by ronnodas** [comment]
It could suggest `foo().as_ref()?` instead.

**Comment by xkr47** [comment]
I have a similar one (afaik did not happen before 1.97.0):

```rust
struct X {
    it: Option<AnotherIterator>,
}

impl Iterator for X {
    type Item = Y;

    fn next(&mut self) -> Option<Self::Item> {
        let it = match &mut self.it {
            Some(it) => it,
            None => return None,
        };
	...
    }
}
```
After fix:
```rust
    let it: &mut AnotherIterator = &mut self.it?;


error[E0507]: cannot move out of `self.it` which is behind a mutable reference
   --> src/foo.rs:249:45
    |
249 |         let it: &mut AnotherIterator = &mut self.it?;
    |                                             ^^^^^^^-
    |                                             |
    |                                             `self.it` moved due to this method call
    |                                             move occurs because `self.it` has type `std::option::Option<AnotherIterator>`, which does not implement the `Copy` trait
```
